x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/workspace_panel.tsx

```diff
@@ -69,19 +70,62 @@ export function WorkspacePanel({
   }
 
   function renderVisualization() {
+    const [expressionError, setExpressionError] = useState(false);
+
+    const activeVisualization = activeVisualizationId && visualizationMap[activeVisualizationId];
+    const expression = useMemo(
+      () =>
+        activeVisualization &&
+        buildExpression(
+          activeVisualization,
+          visualizationState,
+          activeDatasource,
+          datasourceState,
+          datasourcePublicAPI
+        ),
+      [
+        activeVisualization,
+        visualizationState,
+        activeDatasource,
+        datasourceState,
+        datasourcePublicAPI,
+      ]
+    );
+
+    useEffect(
+      () => {
+        // reset expression error if component attempts to run it again
+        if (expressionError) {
+          setExpressionError(false);
+        }
+      },
+      [expression]
+    );
+
     if (activeVisualizationId === null) {
       return renderEmptyWorkspace();
     }
 
-    const activeVisualization = visualizationMap[activeVisualizationId];
-    const datasourceExpression = activeDatasource.toExpression(datasourceState);
-    const visualizationExpression = activeVisualization.toExpression(
-      visualizationState,
-      datasourcePublicAPI
-    );
-    const expression = `${datasourceExpression} | ${visualizationExpression}`;
-
-    return <ExpressionRendererComponent expression={expression} />;
+    if (expression && !expressionError) {
```

Comment:
Since you're already guarding on the expression being truthy, I think it's fine to return any falsy value from `buildExpression` if the data source expression is empty.